ledger-service/lf-value-json/src/main/scala/com/digitalasset/daml/lf/value/json/ApiCodecCompressed.scala

```diff
@@ -97,96 +108,130 @@ abstract class ApiCodecCompressed[Cid](
       value: JsValue,
       prim: Model.DamlLfTypePrim,
       defs: Model.DamlLfTypeLookup): V[Cid] = {
-    (value, prim.typ) match {
-      case (JsString(v), Model.DamlLfPrimType.Decimal) =>
-        V.ValueDecimal(assertDE(LfDecimal fromString v))
-      case (JsString(v), Model.DamlLfPrimType.Int64) => V.ValueInt64(v.toLong)
-      case (JsString(v), Model.DamlLfPrimType.Text) => V.ValueText(v)
-      case (JsString(v), Model.DamlLfPrimType.Party) =>
-        V.ValueParty(assertDE(Ref.Party fromString v))
-      case (v, Model.DamlLfPrimType.ContractId) => V.ValueContractId(jsValueToApiContractId(v))
-      case (JsObject(_), Model.DamlLfPrimType.Unit) => V.ValueUnit
-      case (JsString(v), Model.DamlLfPrimType.Timestamp) => V.ValueTimestamp.fromIso8601(v)
-      case (JsString(v), Model.DamlLfPrimType.Date) => V.ValueDate.fromIso8601(v)
-      case (JsBoolean(v), Model.DamlLfPrimType.Bool) => V.ValueBool(v)
-      case (JsArray(v), Model.DamlLfPrimType.List) =>
-        V.ValueList(v.map(e => jsValueToApiValue(e, prim.typArgs.head, defs)).to[FrontStack])
-      case (JsObject(f), Model.DamlLfPrimType.Optional) =>
-        f.headOption match {
-          case Some((`fieldNone`, _)) => V.ValueOptional(None)
-          case Some((`fieldSome`, v)) =>
-            V.ValueOptional(Some(jsValueToApiValue(v, prim.typArgs.head, defs)))
-          case Some(_) => deserializationError(s"Can't read ${value.prettyPrint} as Optional")
-          case None => deserializationError(s"Can't read ${value.prettyPrint} as Optional")
+    (prim.typ, value).match2 {
```

Comment:
Using `match2` is to preserve exhaustiveness checking? Or is there any other advantage?

Reply:
Yes; additionally, we might decide to resolve the first match layer in advance of having the `value` in the future, notwithstanding DAML recursive cases.